### PR TITLE
Parse (area-specific) context in template-part block.

### PR DESCRIPTION
**Pull request by felixarntz**
## What?
Similar to #45534, this PR is part of the solution from https://github.com/WordPress/wordpress-develop/pull/3560, which addresses https://core.trac.wordpress.org/ticket/56930. We need to pass the `$context` parameter to `wp_filter_content_tags()` in the `template-part` block.

## Why?
We should always pass a context parameter to `wp_filter_content_tags()`, except when there is a current filter running. This is not the case when parsing the block template though, which is why https://github.com/WordPress/wordpress-develop/pull/3560 already takes care of it for the overall template. It also does so for the `template-part` block, however since that can only be edited within upstream Gutenberg, I have opened this PR.

Including the template part area identifier is useful context for that function, as it can make different decisions based on that. For example, images in the header of a site are extremely likely to appear above the fold, while for images in the footer the opposite applies.

## How?
The implementation here goes a bit beyond the core PR https://github.com/WordPress/wordpress-develop/pull/3560, using a cleaner implementation: We should ensure that the `$area` string is one of the allowed area identifiers before passing it through in the context.

## Testing Instructions
Test this as part of https://github.com/WordPress/wordpress-develop/pull/3560:

1. Add a `header` template part with an image in it.
2. Add a `footer` template part with an image in it.
3. See how the image in the header is not lazy-loaded, while the one in the footer is lazy-loaded.

### CODE REVIEW

**Comment by felixarntz** [comment]
@Mamaduka Can you take a look at this?

Apologies for asking that again (especially after we now already backported a few package updates in https://core.trac.wordpress.org/changeset/55079 🤦‍♂️), but if this is good to go, is there any chance that we can backport it soon? Unfortunately it blocks the same core PR https://github.com/WordPress/wordpress-develop/pull/3560. To be fair, since we're not far from 6.2 beta, can you tell me when the regular package update would happen for core? Maybe that would be enough.

Anyway, first of all please take a look at the PR :)

**Comment by carlomanf** [comment]
Why use the prefix `the_`? Introducing historical conventions from b2/cafelog into block themes seems very backward. Some care should be taken to choose the right string, as it probably can't ever be changed.

File: packages/block-library/src/template-part/index.php [hunk 1]
```diff
@@ -126,27 +126,39 @@ function render_block_core_template_part( $attributes ) {
...
+
 	// Run through the actions that are typically taken on the_content.
 	$seen_ids[ $template_part_id ] = true;
 	$content                       = do_blocks( $content );
 	unset( $seen_ids[ $template_part_id ] );
 	$content = wptexturize( $content );
 	$content = convert_smilies( $content );
 	$content = shortcode_unautop( $content );
-	$content = wp_filter_content_tags( $content );
+	$content = wp_filter_content_tags( $content, "the_template_part_{$area}" );
```
Comment (carlomanf) [inline packages/block-library/src/template-part/index.php]: Why use the prefix `the_`? Introducing historical conventions from b2/cafelog into block themes seems very backward. Some care should be taken to choose the right string, as it probably can't ever be changed.

Comment (felixarntz) [reply packages/block-library/src/template-part/index.php]: @carlomanf I agree with you on the one hand, one the other hand the values used as `$context` in core so far are those older ones based on filters, e.g. `the_content`, `the_post_thumbnail`. I chose this string to align with that.

But if you feel strongly about getting rid of the `the_` prefix despite other context values using it, I'm certainly open to doing that.

Comment (carlomanf) [reply packages/block-library/src/template-part/index.php]: Not all of the context values that currently pass through `wp_filter_content_tags` start with `the_`, only the ones based on b2-style template tags. `widget_text_content` and `widget_block_content` also pass through `wp_filter_content_tags` so I think there is full freedom to choose a name that describes the context well, i.e. that it's a block callback.

You may want to look at WordPress/wordpress-develop#3851 and whether the function call there should use the same value.

**Review by Mamaduka — COMMENTED** [review]
Thanks for the ping, @felixarntz.

> To be fair, since we're not far from 6.2 beta, can you tell me when the regular package update would happen for core? Maybe that would be enough.

Technically we can start doing regular package releases for the 6.2 beta after Gutenberg 15.1 RC is released. It's scheduled for next week. I can ensure we merge this update before RC so it's included in the initial package updates PR.

We also have a similar issue with the [Cover block](https://github.com/WordPress/gutenberg/issues/46789) when it's used as a featured image. It might be a good idea to patch it as well.

File: packages/block-library/src/template-part/index.php (hunk 1 above)
Comment (felixarntz) [reply packages/block-library/src/template-part/index.php]: @carlomanf Fair point, will update. 👍

**Comment by felixarntz** [comment]
@Mamaduka I reviewed the issue, tested it, and left a reply. Please take a look.

For context, the core ticket https://core.trac.wordpress.org/ticket/56930 already covers the part of the issues reported that is realistic to cover at this point. There is one issue there that remains (`core/cover` usage in block templates), however fixing that is out of scope for that ticket, since that usage would be outside of "the loop". Any usage outside of the loop is also a problem for classic themes FWIW. It is something that members of the performance team including myself are actively looking into fixing, however it will require another separate workstream and ticket. 56930 will at least fix the most prevalent problems, for when such images are used within the loop.

In other words, this PR should be good as is. It would be great if we could get it into the Gutenberg 15.1 RC release.

**Comment by Mamaduka** [comment]
Thank you, @felixarntz!

I'll make sure to get this merged before 15.1 RC, more likely before the end of the week.

**Review by carlomanf — APPROVED** [review]

**Comment by felixarntz** [comment]
@Mamaduka Friendly ping here. Just want to make sure this isn't forgotten :)

**Review by Mamaduka — APPROVED** [review]
This looks great.

I believe https://github.com/WordPress/wordpress-develop/pull/3560 will need a refresh to use the updated context name.
